File: 21-Binary-Search-Tree-Part-II/python/07_maximum_sum_bst_in_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        self.ans = 0
        inf = 1 << 30

        def dfs(node: Optional[TreeNode]):
            if not node:
                return (1, inf, -inf, 0)

            l_is_bst, l_min, l_max, l_sum = dfs(node.left)
            r_is_bst, r_min, r_max, r_sum = dfs(node.right)
            v = node.val

            if l_is_bst and r_is_bst and l_max < v < r_min:
                s = v + l_sum + r_sum
                self.ans = max(self.ans, s)
                return (1, min(l_min, v), max(r_max, v), s)

            return (0, 0, 0, 0)

        dfs(root)
        return self.ans

File: 21-Binary-Search-Tree-Part-II/cpp/07-maximum-sum-bst-in-binary-tree.cpp
#include <vector>
#include <algorithm>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans = 0;
    const int inf = 1 << 30;

    // returns {isBST, minVal, maxVal, sum}
    vector<int> dfs(TreeNode* root) {
        if (!root) {
            return {1, inf, -inf, 0};
        }
        auto l = dfs(root->left);
        auto r = dfs(root->right);
        int v = root->val;

        if (l[0] == 1 && r[0] == 1 && l[2] < v && r[1] > v) {
            int s = v + l[3] + r[3];
            ans = max(ans, s);
            return {1, min(l[1], v), max(r[2], v), s};
        }
        return {0, 0, 0, 0};
    }

public:
    int maxSumBST(TreeNode* root) {
        dfs(root);
        return ans;
    }
};
